consensus/transactions.go

```diff
@@ -53,39 +58,152 @@ func (s *State) validInput(input SiacoinInput) (err error) {
-// returns an error explaining what wasn't valid.
+// validSiafundInput checks that the given input is valid within the current
+// consensus state. If not, an error is returned.
+func (s *State) validSiafundInput(sfi SiafundInput) (sfo SiafundOutput, err error) {
+	// Check the input spends an existing and valid output.
+	sfo, exists := s.siafundOutputs[sfi.ParentID]
+	if !exists {
+		err = ErrMissingSiafundOutput
+		return
+	}
+
+	// Check that the spend conditions match the hash listed in the output.
+	if sfi.UnlockConditions.UnlockHash() != sfo.UnlockHash {
+		err = errors.New("spend conditions do not match hash")
+		return
+	}
+
+	// Check the timelock on the spend conditions is expired.
+	if sfi.UnlockConditions.Timelock > s.height() {
+		err = errors.New("output spent before timelock expiry.")
+		return
+	}
+
+	return
+}
+
+func (s *State) validFileContractTermination(fct FileContractTermination) (err error) {
+	// Check that the FileContractTermination terminates an existing
+	// FileContract.
+	fc, exists := s.fileContracts[fct.ParentID]
+	if !exists {
+		err = ErrMissingFileContract
+		return
+	}
+
+	// Check that the spend conditions match the hash listed in the output.
+	if fct.TerminationConditions.UnlockHash() != fc.TerminationHash {
+		err = errors.New("spend conditions do not match hash")
+		return
+	}
+
+	// Check the timelock on the spend conditions is expired.
+	if fct.TerminationConditions.Timelock > s.height() {
+		err = errors.New("contract terminated before timelock expiry.")
+		return
+	}
+
+	// Check that the payouts in the termination add up to the payout of the
+	// contract.
+	var payoutSum Currency
+	for _, payout := range fct.Payouts {
+		err = payoutSum.Add(payout.Value)
+		if err != nil {
+			return
+		}
+	}
+	if payoutSum.Cmp(fc.Payout) != 0 {
+		err = errors.New("contract termination has incorrect payouts")
+		return
+	}
+
+	return
+}
+
+// validSiafunds checks that the transaction has valid siafund inputs and
+// outputs, and that the sum of the inputs matches the sum of the outputs.
+func (s *State) validSiafunds(t Transaction) (err error) {
+	// Check that all siafund inputs are valid, and get the total number of
+	// input siafunds.
+	var siafundInputSum Currency
+	for _, sfi := range t.SiafundInputs {
+		// Check that the input is valid.
+		var sfo SiafundOutput
+		sfo, err = s.validSiafundInput(sfi)
+		if err != nil {
+			return
+		}
+
+		// Add this input's value
+		err = siafundInputSum.Add(sfo.Value)
+		if err != nil {
+			return
+		}
+	}
+
+	// Check that all siafund outputs are valid and that the siafund output sum
+	// is equal to the siafund input sum.
+	var siafundOutputSum Currency
+	for _, sfo := range t.SiafundOutputs {
+		// Check that the claimStart is set to 0. Type safety should enforce
+		// this, but check anyway.
+		if sfo.ClaimStart.Cmp(ZeroCurrency) != 0 {
+			return errors.New("invalid siafund output presented")
+		}
```

Comment:
I believe there's an `IsZero` method for this purpose. What do you mean about type safety?